test:(tool): A tool for batch sending and checking test transactions.

TimmyExogenous opened the pull request:
## Description

  *  Implement object creation and funding.
  *  Implement the registration of assets, stakers, operators, and AVSs.
  *  Implement a function to opt all operators into all AVSs for testing.
  *  Implement the batch sending for deposit, delegation, undelegation and withdrawal transactions
  *  Implement the batch check for deposit, withdrawal, delegation and undelegation txs

## todo

 * the entire process
 * command line
 * test 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Updated comments to reflect the correct context for the `BaseDenom` constant.
	- Introduced new dependencies for enhanced functionality.
	- Added bindings for Ethereum smart contracts related to asset management, AVS, and delegation.
	- Implemented a transaction management system for batch processing.
	- Enhanced command-line tool for testing and transaction management with new commands and functionalities.

- **Bug Fixes**
	- Improved error handling in various methods to ensure robust operation.

- **Documentation**
	- Updated README.md for the `exocore-test-tool` to clarify its functionalities and usage.

- **Tests**
	- Added comprehensive test cases for the batch package to ensure functionality and reliability.

- **Chores**
	- Introduced new Makefile targets for building and installing the test tool.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TimmyExogenous committed:
1.Add config, type.go, and manager for the E2E tool.
2.Implement object creation and funding.
3.Implement the registration of assets, stakers, operators, and AVSs.
4.Implement a function to opt all operators into all AVSs for testing.
5.Implement the batch sending for deposit, delegation, undelegation and withdrawal transactions
6.Implement the batch check for deposit, withdrawal, delegation and undelegation txs

TimmyExogenous committed:
add fetchAndSaveDogfoodAVS

TimmyExogenous committed:
1. add the entire process of automatical sending and  checking for batch test transactions.
2. add the command-line for the test tool

TimmyExogenous committed:
add test for debug and fix the issues identified by local test

TimmyExogenous committed:
Add preparation check and Cosmos transaction waiting function.

TimmyExogenous committed:
fix the issue regarding sqlite db lock conflict
set wait duration for sqlite lock conflict
fixed the issue regarding batch opting in
fixed the issue regarding object operation and registration

TimmyExogenous committed:
add more commands fot the test tool

TimmyExogenous committed:
rebase to develop and fix the issues in go.mod and command line

TimmyExogenous commented on testutil/batch/utils.go:
```diff
@@ -0,0 +1,175 @@
+		return err
+	}
+	return nil
+}
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,130 @@
+		WithGasAdjustment(GasAdjustment).
+		WithSimulateAndExecute(true)
+	clientCtx.WithFrom(fromName)
+	clientCtx.WithBroadcastMode(broadcastMode)
```
it has been fixed.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,130 @@
+	receipt, err := WaitForEvmTxReceipt(basicInfo.ethC, txHash, basicInfo.WaitDuration)
+	if err != nil {
+		return err
+	}
```
outdated

TimmyExogenous commented on testutil/batch/utils.go:
```diff
@@ -0,0 +1,175 @@
+	err := m.GetDB().AutoMigrate(&model)
+	if err != nil {
+		return xerrors.Errorf("Auto migration failed for %T table, err: %s", model, err)
+	}
```
It's only used as a test tool. So it's fine.

TimmyExogenous commented on testutil/batch/utils.go:
```diff
@@ -0,0 +1,175 @@
+	// Automatically migrate the schema, creating the table if it doesn't exist
+	if err := m.GetDB().AutoMigrate(&obj); err != nil {
+		return xerrors.Errorf("Failed to auto migrate schema for %T, err: %s", obj, err)
+	}
```
It's only used as a test tool. So it's fine.

TimmyExogenous commented on testutil/batch/utils.go:
```diff
@@ -0,0 +1,175 @@
+	return nil
+}
+
+func PaddingAddressTo32(address common.Address) []byte {
+	paddingLen := 32 - len(address)
+	ret := make([]byte, len(address))
+	copy(ret, address[:])
+	for i := 0; i < paddingLen; i++ {
+		ret = append(ret, 0)
+	}
+	return ret
+}
```
It's fine to append zeros to the address, the current application scenario is designed this way.

TimmyExogenous commented on testutil/batch/utils.go:
```diff
@@ -0,0 +1,175 @@
+		}
+	}
+	return nil
+}
```
Thank you for the reminder. This is only used in a test tool, and we won't delete the test objects, so non-sequential IDs won't be an issue in this case.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,173 @@
+	Sk           []byte         // private key
+	Transactions []Transaction  `gorm:"foreignKey:StakerID"`
+	// AvailableNonce int64
+}
```
The private keys are only used for testing, so it's fine.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,173 @@
+	ID      uint           `gorm:"primaryKey"`                   // primary key
+	Address common.Address `gorm:"column:address;type:char(42)"` // eth address
+	Sk      []byte         // private key
+}
```
The private keys are only used for testing, so it's fine.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,173 @@
+	StakerNumber   int `mapstructure:"staker-number"`
+	OperatorNumber int `mapstructure:"operator-number"`
+	AVSNumber      int `mapstructure:"avs-number"`
+	AsssetNumber   int `mapstructure:"asset-number"`
```
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,173 @@
+	OnChainAndSuccessful        // On-chain and successful
+)
+
+const (
+	WaitToCheck = iota
+	Successful
+	failed
+)
```
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+func (o Operator) EvmAddress() common.Address {
+	accAddr, _ := sdktypes.AccAddressFromBech32(o.Address)
+	return common.BytesToAddress(accAddr[:common.AddressLength])
+}
```
The operator address is generated automatically, so it's fine not to address this error.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+func (o Operator) AccAddress() sdktypes.AccAddress {
+	accAddr, _ := sdktypes.AccAddressFromBech32(o.Address)
+	return accAddr
```
The operator address is generated automatically, so it's fine not to address this error.

TimmyExogenous commented on testutil/batch/tx_check.go:
```diff
@@ -0,0 +1,228 @@
+	}
+
+	if err != nil {
+		return nil, 0, xerrors.Errorf("Failed to retrieve transaction IDs with TestBatchID %d and Type %s, err: %s", batchID, txType, err)
```
Thanks, I will still use the xerrors package but change the placeholder from `%v` to `%w`.

TimmyExogenous commented on testutil/batch/tx_check.go:
```diff
@@ -0,0 +1,228 @@
+	delegationtype "github.com/ExocoreNetwork/exocore/x/delegation/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"golang.org/x/xerrors"
+)
```
It's fine to use the xerrors package, although Go 1.13 provides error wrapping because it won't capture a stack backtrace.

TimmyExogenous commented on testutil/batch/tx_check.go:
```diff
@@ -0,0 +1,228 @@
+				if res.UndelegatableShare.Equal(expectedAmount) {
+					transaction.CheckResult = Successful
+				} else {
+					transaction.CheckResult = failed
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,284 @@
+		// lead to failures in the deposit and withdrawal checks.
+		singleDelegationAmount := big.NewInt(0).Quo(totalDelegationAmount, big.NewInt(operatorNumber))
+
+		assetOpFunc := func(assetId uint, _ int64, asset *Asset) error {
```
outdated

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,284 @@
+		return false, nil
+	}
+	return m.TickHandle(m.config.TxNumberPerSec, handle)
+}
```
Thanks for the reminder. We might add unit tests when we have some time. Since it's just a test tool, the priority for unit tests is low for now.

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,284 @@
+			return err
+		}
+		txID = txHash.String()
+	}
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	// Calculate the interval time in milliseconds for each transaction based on the rate
+	interval := time.Duration(1000/handleRate) * time.Millisecond
+	ticker := time.NewTicker(interval)
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,284 @@
+				data, err := delegationAbi.Pack(msgType, asset.ClientChainID, nonce, PaddingAddressTo32(asset.Address), PaddingAddressTo32(staker.Address), []byte(operator.Address), singleDelegationAmount)
+				if err != nil {
+					return err
+				}
```
we use the nonce of the caller as the LayerZero nonce, So it's fine here.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,269 @@
+}
+
+func (m *Manager) Close() {
+	for _, client := range m.NodeEVMHTTPClients {
+		client.Close()
+	}
+	for _, client := range m.NodeEVMWSClients {
+		client.Close()
+	}
+}
```
They have a `close` function, which will call the `close` method of the RPC client.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,269 @@
+	// creat client context for the nodes
+	for _, url := range config.NodesRPC {
+		// Create a client context to connect to the Cosmos node
+		clientCtx := client.Context{}.
+			WithNodeURI(url).            // gRPC address of the Cosmos node
+			WithChainID(config.ChainID). // Chain ID of the Cosmos blockchain
+			WithKeyring(KeyRing)
+		manager.NodeClientCtx = append(manager.NodeClientCtx, clientCtx)
+	}
+
+	evmChainID, err := manager.NodeEVMHTTPClients[DefaultNodeIndex].ChainID(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("can't get the evm chainID, err:%s", err)
+	}
+	ethSigner := ethtypes.LatestSignerForChainID(evmChainID)
+	manager.DefaultEvmTxRequirements = &BasicEvmTxRequirements{
+		ctx:          manager.ctx,
+		sk:           manager.FaucetSK,
+		caller:       crypto.PubkeyToAddress(sk.PublicKey),
+		signer:       ethSigner,
+		ethC:         manager.NodeEVMHTTPClients[DefaultNodeIndex],
+		WaitDuration: time.Duration(config.TxCheckInterval),
+	}
+	return manager, nil
+}
+
+func (m *Manager) Close() {
+	for _, client := range m.NodeEVMHTTPClients {
+		client.Close()
+	}
+	for _, client := range m.NodeEVMWSClients {
+		client.Close()
+	}
+}
+
+func (m *Manager) GetDB() *gorm.DB {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return m.db
+}
```
Thanks, the lock will be removed in the next commit.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,269 @@
+		db:                 db,
+		FaucetSK:           sk,
+		KeyRing:            KeyRing,
+		NodeEVMHTTPClients: make([]*ethclient.Client, config.ChainValidatorNumber),
+		NodeEVMWSClients:   make([]*ethclient.Client, config.ChainValidatorNumber),
+		NodeClientCtx:      make([]client.Context, config.ChainValidatorNumber),
+		TxsQueue:           make(chan interface{}, config.TxsQueueBufferSize),
```
thanks, the slice will be initialized through the index. so it's fine here.

TimmyExogenous commented on testutil/batch/prepare.go:
```diff
@@ -0,0 +1,208 @@
+		req := &assetstypes.QueryStakingAssetInfo{
+			AssetID: assetID, // already lowercase
+		}
+		_, err := queryClient.QueStakingAssetInfo(m.ctx, req)
```
It's not a typo,

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,333 @@
+				// withdraw all amount
+				opAmount = stakerAssetInfo.WithdrawableAmount
+				expectedCheckValue = expectedCheckValue.Sub(opAmount)
+			}
```
this check is unnecessary, because it's used for the test tool

TimmyExogenous commented on testutil/batch/tx_check.go:
```diff
@@ -0,0 +1,277 @@
+				transaction.ActualCheckValue = res.TotalDepositAmount.String()
+			}
+			return nil
+		}
```
This function isn't complex, so it's fine not to refactor as you suggested.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,313 @@
+	if nonceFromSyncMap {
+		// update the sequence if we use faucet sk to sign the tx
+		m.Sequences.Store(crypto.PubkeyToAddress(m.FaucetSK.PublicKey), nonce+1)
+	}
```
outdated.

TimmyExogenous commented on testutil/batch/manager_test.go:
```diff
@@ -0,0 +1,120 @@
+//go:build skiptests
+
```
This test file is used for debugging, it depends on an Exocore chain node. So we need to add this flag to ignore this test file.

TimmyExogenous commented on Makefile:
```diff
@@ -136,6 +136,11 @@ $(BUILD_TARGETS): go.sum $(BUILDDIR)/
 $(BUILDDIR)/:
 	mkdir -p $(BUILDDIR)/
 
+build-test-tool:
+	CGO_ENABLED="1" go build $(BUILD_FLAGS) $(BUILD_ARGS) ./cmd/test-tool
+install-test-tool:
+	CGO_ENABLED="1" go install $(BUILD_FLAGS) $(BUILD_ARGS) ./cmd/test-tool
```
thanks, it will be removed in the next commit.

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,344 @@
+	if err != nil {
+		return err
+	}
+	evmTx.TxRecord.SendTime = time.Now().String()
```
It's fine here because it's only used to record the test information for the test tool. It won't be part of the consensus.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,568 @@
+		return err
+	}
+	// send test transactions in batch
+	go func() {
+		// deposit
+		if err := m.EnqueueAndCheckTxsInBatch(assets.MethodDepositLST); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all deposit txs")
+		// delegation
+		if err := m.EnqueueAndCheckTxsInBatch(delegation.MethodDelegate); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all delegation txs")
+		// undelegation
+		if err := m.EnqueueAndCheckTxsInBatch(delegation.MethodUndelegate); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all undelegation txs")
+		// withdrawal
+		if err := m.EnqueueAndCheckTxsInBatch(assets.MethodWithdrawLST); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all withdrawal txs")
+		time.Sleep(time.Duration(m.config.EachTestInterval) * time.Second)
+	}()
```
It's fine here because it's only used for the test tool.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,308 @@
+
+// WaitForEvmTxReceipt Wait for the transaction receipt and confirm if it's mined successfully
+func WaitForEvmTxReceipt(client *ethclient.Client, txHash common.Hash, waitDuration, waitExpiration time.Duration) (*types.Receipt, error) {
+	startTime := time.Now() // Record the start time of the function
```
It's fine here because it's only used for the test tool.

TimmyExogenous commented on local_node.sh:
```diff
@@ -312,4 +312,4 @@ EOF
 
 # Start the node (remove the --pruning=nothing flag if historical queries are not needed)
-exocored start --metrics "$TRACE" --log_level $LOGLEVEL --minimum-gas-prices=0.0001hua --json-rpc.api eth,txpool,personal,net,debug,web3 --api.enable --json-rpc.enable true --home "$HOMEDIR" --chain-id "$CHAINID" --oracle --grpc.enable true
+exocored start --metrics "$TRACE" --log_level $LOGLEVEL --minimum-gas-prices=0.0001hua --json-rpc.api eth,txpool,personal,net,debug,web3 --api.enable --json-rpc.enable true --home "$HOMEDIR" --chain-id "$CHAINID" --grpc.enable true
```
thanks, It has been readded.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,128 @@
+		file, err := os.Create(configFilePath)
+		if err != nil {
+			fmt.Printf("failed to create config file: %s", err)
+		}
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,128 @@
+		if err != nil {
+			fmt.Printf("failed to create config file: %s", err)
+		}
+		defer file.Close()
```
it will be fixed in the next commit. The function will return if it fails to create the config file.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,568 @@
+	if err != nil {
+		return nil, xerrors.Errorf("can't new a test key ring, err:%s", err)
+	}
+	_, err = KeyRing.Key(FaucetSKName)
+	if err != nil {
+		err = KeyRing.ImportPrivKeyHex(FaucetSKName, config.FaucetSk, ethsecp256k1.KeyType)
+		if err != nil {
+			return nil, xerrors.Errorf("can't import the faucet private key, err:%s", err)
+		}
+	}
```
Thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,244 @@
+	ID           uint `gorm:"primaryKey"` // primary key
+	Name         string
+	Address      common.Address `gorm:"column:address;type:char(42)"` // eth address
+	Sk           []byte         // private key
```
It's fine here because all privates are generated automatically and are only used for the test tool.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,244 @@
+	ActualCheckValue   string
+}
+
+type Operator struct {
+	ID              uint `gorm:"primaryKey"` // primary key
+	Name            string
+	Address         string // cosmos address
+	Sk              []byte // private key
+	ConsensusPubKey string
+	ConsensusSk     string
+}
```
It's fine here because all privates are generated automatically and are only used for the test tool.

TimmyExogenous commented on testutil/batch/type.go:
```diff
@@ -0,0 +1,244 @@
+	ID      uint `gorm:"primaryKey"` // primary key
+	Name    string
+	Address string // eth address
+	Sk      []byte // private key
```
It's fine here because all privates are generated automatically and are only used for the test tool.

TimmyExogenous commented on testutil/batch/tx_check.go:
```diff
@@ -0,0 +1,282 @@
+				logger.Info("DepositWithdrawLSTCheck, the check is failed", "txID", transaction.TxHash,
+					"staker", staker.Name, "stakerAddr", staker.EvmAddress(),
+					"asset", asset.Name, "assetAddr", asset.Address)
+				transaction.CheckResult = failed
```
It will be fixed in the next commit.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,306 @@
+	Run: func(_ *cobra.Command, args []string) {
+		batchID, err := strconv.ParseUint(args[1], 10, 32)
+		if err != nil {
+			fmt.Printf("invalid batch id,input:%s,err:%s", args[0], err)
+			return
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,306 @@
+			fmt.Printf("invalid status,input:%s,err:%s", args[1], err)
+			return
+		}
+		if int(status) > batch.OnChainAndSuccessful {
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,306 @@
+			fmt.Printf("invalid status,status:%d", status)
+			return
+		}
+		txIDs, count, err := batch.GetTxIDsByBatchTypeAndStatus(sqliteDB, uint(batchID), args[0], int(status))
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,308 @@
+		}
+
+		// Attempt to retrieve the transaction receipt
+		receipt, err := client.TransactionReceipt(context.Background(), txHash)
```
It will be updated in the next commit.

TimmyExogenous commented on testutil/batch/manager_test.go:
```diff
@@ -0,0 +1,209 @@
+
+func Test_LoadAllAVSs(t *testing.T) {
+	// open the sqlite db
+	homePath := "/home/timmy/tests/test-tool"
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/manager_test.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+func Test_FaucetSk(t *testing.T) {
+	appManager, err := NewManager(context.Background(), "/home/timmy/tests/test-tool", &DefaultTestToolConfig)
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/manager_test.go:
```diff
@@ -0,0 +1,209 @@
+
+func Test_LoadTxRecords(t *testing.T) {
+	// open the sqlite db
+	homePath := "/home/timmy/tests/test-tool"
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/manager_test.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+func Test_QueryAllBalance(t *testing.T) {
+	appManager, err := NewManager(context.Background(), "/home/timmy/tests/test-tool", &DefaultTestToolConfig)
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,306 @@
+			fmt.Printf("invalid status,input:%s,err:%s", args[1], err)
+			return
+		}
+		if int(status) > batch.OnChainAndSuccessful {
```
it will be fixed in the next commit.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,306 @@
+			fmt.Printf("invalid status,status:%d", status)
+			return
+		}
+		txIDs, count, err := batch.GetTxIDsByBatchTypeAndStatus(sqliteDB, uint(batchID), args[0], int(status))
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,612 @@
+	return nil
+}
+
+func (m *Manager) CreateOperators() error {
+	createNewOperator := func(id uint) (Operator, error) {
+		addr, privKey := testutiltx.NewAccAddressAndKey()
+		name := fmt.Sprintf("%s%d", OperatorNamePrefix, id)
+		// add the Sk to key ring
+		err := m.KeyRing.ImportPrivKeyHex(name, common.Bytes2Hex(privKey.Key), ethsecp256k1.KeyType)
+		if err != nil {
+			return Operator{}, err
+		}
+
+		privVal := mock.NewPV()
+		pubKey, err := privVal.GetPubKey()
+		if err != nil {
+			return Operator{}, err
+		}
+		consensusKey := keytypes.NewWrappedConsKeyFromHex(hexutil.Encode(pubKey.Bytes()))
+		return Operator{
+			Name:            name,
+			Address:         strings.ToLower(addr.String()),
+			Sk:              privKey.Bytes(),
+			ConsensusPubKey: consensusKey.ToJSON(),
+			ConsensusSk:     privVal.PrivKey.String(),
+		}, nil
+	}
+	// create a new Staker
+	return CreateObjects(m.GetDB(), Operator{}, int64(m.config.OperatorNumber), createNewOperator)
+}
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on precompiles/assets/assets_binding.go:
```diff
@@ -0,0 +1,396 @@
+//
+// Solidity: function getClientChains() view returns(bool, uint32[])
+func (_Assets *AssetsCaller) GetClientChains(opts *bind.CallOpts) (bool, []uint32, error) {
+	var out []interface{}
+	err := _Assets.contract.Call(opts, &out, "getClientChains")
+
+	if err != nil {
+		return *new(bool), *new([]uint32), err
+	}
+
+	out0 := *abi.ConvertType(out[0], new(bool)).(*bool)
+	out1 := *abi.ConvertType(out[1], new([]uint32)).(*[]uint32)
+
+	return out0, out1, err
+
```
This file is generated by `abigen`, so it's fine here.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,612 @@
+	}
+
+	// send test transactions in batch
+	go func() {
+		// deposit
+		if err := m.EnqueueAndCheckTxsInBatch(assets.MethodDepositLST); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all deposit txs")
+		// delegation
+		if err := m.EnqueueAndCheckTxsInBatch(delegation.MethodDelegate); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all delegation txs")
+		// undelegation
+		if err := m.EnqueueAndCheckTxsInBatch(delegation.MethodUndelegate); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all undelegation txs")
+		// withdrawal
+		if err := m.EnqueueAndCheckTxsInBatch(assets.MethodWithdrawLST); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("Start: finish enqueuing and checking all withdrawal txs")
+		time.Sleep(time.Duration(m.config.EachTestInterval) * time.Second)
+	}()
```
Seems unnecessary, because it's only used for the test tool. And the next call `DequeueAndSignSendTxs` will block the process. So how to use `eg.Wait`

TimmyExogenous replied:
Thanks, I will change it in the next commit.

TimmyExogenous committed:
fix the lint error and issues from code comments

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,610 @@
+		txsCheckFunc = m.EvmDelegationCheck
+	default:
+		return xerrors.Errorf("EnqueueAndCheckTxsInBatch, invalid msgType:%s", msgType)
+	}
+	logger.Info("EnqueueAndCheckTxsInBatch: call enqueueTxsFunc", "msgType", msgType, "batchID", helperRecord.CurrentBatchID)
+	if err := enqueueTxsFunc(helperRecord.CurrentBatchID, msgType); err != nil {
+		logger.Error("EnqueueAndCheckTxsInBatch: Failed to test transactions in batch", "msgType", msgType, "err", err)
+		return err
+	}
+	// When configuring this interval, the durations of sending and on-chain processing should be taken into account.
+	time.Sleep(time.Duration(m.config.BatchTxsCheckInterval) * time.Second)
+	// Check if all test transactions have been dequeued; if not, wait for them to be dequeued.
+	for {
+		if len(m.TxsQueue) != 0 {
+			time.Sleep(time.Duration(m.config.BatchTxsCheckInterval) * time.Second)
+		} else {
+			break
+		}
+	}
+	logger.Info("EnqueueAndCheckTxsInBatch: call PrecompileTxOnChainCheck", "msgType", msgType, "batchID", helperRecord.CurrentBatchID)
+	if err := m.PrecompileTxOnChainCheck(helperRecord.CurrentBatchID, msgType); err != nil {
+		logger.Error("EnqueueAndCheckTxsInBatch: Failed to check whether the txs are on chain", "err", err)
+		return err
+	}
+	logger.Info("EnqueueAndCheckTxsInBatch: call txsCheckFunc", "msgType", msgType)
+	if err := txsCheckFunc(helperRecord.CurrentBatchID, msgType); err != nil {
+		logger.Error("EnqueueAndCheckTxsInBatch: Failed to check the txs", "msgType", msgType, "err", err)
+		return err
+	}
+	// increase the batch id if the msg type is withdrawal
+	if msgType == assets.MethodWithdrawLST {
+		helperRecord.CurrentBatchID++
+		if err := SaveObject(m.GetDB(), helperRecord); err != nil {
+			logger.Error("EnqueueAndCheckTxsInBatch: can't save the helper record")
+			return err
+		}
+		logger.Info("EnqueueAndCheckTxsInBatch update the batchID", "batchID", helperRecord.CurrentBatchID)
+	}
+	return nil
+}
```
TxsQueue is a buffered channel. How can I avoid conflicts when checking its length?

TimmyExogenous replied:
And using `len(channel)` won't lead to a race condition,

TimmyExogenous replied:
Thanks, it will be fixed in the next commit.

TimmyExogenous committed:
add readme for the test tool

TimmyExogenous commented on testutil/batch/batch_tx.go:
```diff
@@ -0,0 +1,350 @@
+
+import (
+	"math/big"
+	"reflect"
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,610 @@
+		return err
+	}
+	// send test transactions in batch
+	go func() {
+		// enqueue the txs and check them in batch
+		if err := m.EnqueueAndCheckTxsInBatch(msgType); err != nil {
+			close(m.Shutdown)
+			return
+		}
+		logger.Info("ExecuteBatchTestForType: finish enqueuing and checking all txs")
+	}()
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,306 @@
+	Example: "test-tool batch-test depositLST --home .",
+	Args:    cobra.ExactArgs(1),
+	Run: func(_ *cobra.Command, args []string) {
+		// Start the app manager in a separate goroutine
+		go func() {
+			if err := appManager.ExecuteBatchTestForType(args[0]); err != nil {
+				fmt.Printf("failed to execute the batch test for the specified message type,"+
+					"msgType:%s,err: %s\r\n", args[0], err)
+			}
+		}()
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,306 @@
+	Example: "test-tool start --home .",
+	Args:    cobra.NoArgs,
+	Run: func(_ *cobra.Command, _ []string) {
+		// Start the app manager in a separate goroutine
+		go func() {
+			if err := appManager.Start(); err != nil {
+				fmt.Printf("Error starting the test tool: %s\r\n", err)
+			}
+		}()
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,308 @@
+		Nonce:    nonce,
+		To:       msg.To,
+		Value:    msg.Value,
+		Gas:      uint64(math.Round(float64(estimateGas) * GasAdjustment)),
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,308 @@
+	})
+	logger.Info("SignAndSendEvmTx", "from", txInfo.From, "to", retTx.To(),
+		"nonce", retTx.Nonce(), "value", retTx.Value(), "gasPrice", retTx.GasPrice(),
+		"gas", retTx.Gas(), "dataLength", len(retTx.Data()), "time", time.Now().String())
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,308 @@
+
+func (m *Manager) WaitForCosmosTxs(responseList []*sdktypes.TxResponse, waitDuration, waitExpiration time.Duration) error {
+	for _, res := range responseList {
+		startTime := time.Now() // Record the start time of the function
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on testutil/batch/utils.go:
```diff
@@ -0,0 +1,262 @@
+
+import (
+	"math/big"
+	"reflect"
```
It's used in the test tool, so please dismiss it.

TimmyExogenous commented on cmd/test-tool/README.md:
```diff
@@ -0,0 +1,58 @@
+
+## commands
+
+* The `init` command can be used to create a default configuration file.
+* The `start` command provides all the testing functionalities mentioned above.
+* The `query` commands are used to retrieve information about test objects and the status of test transactions.
+* The `prepare` and `batch-test` commands break down the functionalities provided by the start command. The `prepare`
+  command handles the creation, funding, registration, and opting-in of test objects, while the `batch-test` command
+  allows for manual batch testing of different transaction types.
```
These details can be obtained by `--help`. So we don't add them in README currently.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,612 @@
+		logger.Info("websocket url", "url", url)
+		rc, err := rpc.DialContext(context.Background(), url)
+		if err != nil {
+			return nil, xerrors.Errorf("can't creat the evm websocket rpc, err:%w, url:%s", err, url)
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,612 @@
+		logger.Info("http url", "url", url)
+		rc, err := rpc.DialContext(context.Background(), url)
+		if err != nil {
+			return nil, xerrors.Errorf("can't creat the evm http rpc, err:%w, url:%s", err, url)
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,612 @@
+			return nil, xerrors.Errorf("too many http rpc,index:%d,nodeNumber:%d", i, config.ChainValidatorNumber)
+		}
+		logger.Info("http url", "url", url)
+		rc, err := rpc.DialContext(context.Background(), url)
+		if err != nil {
+			return nil, xerrors.Errorf("can't creat the evm http rpc, err:%w, url:%s", err, url)
+		}
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on testutil/batch/manager.go:
```diff
@@ -0,0 +1,612 @@
+		}
+		logger.Info("websocket url", "url", url)
+		rc, err := rpc.DialContext(context.Background(), url)
+		if err != nil {
+			return nil, xerrors.Errorf("can't creat the evm websocket rpc, err:%w, url:%s", err, url)
+		}
+		c := ethclient.NewClient(rc)
```
thanks, it will be fixed in the next commit.

TimmyExogenous committed:
rebase to the develop branch and fix the conflict
fix some typo and refine the input parameter regarding context

TimmyExogenous committed:
fix the issue in EnqueueDepositWithdrawLSTTxs

bwhour commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,316 @@
+
+// Root command
+var rootCmd = &cobra.Command{
+	Use:   "app",
```
exocore-test-tool

TimmyExogenous replied:
it will be renamed in the next commit.

bwhour commented on cmd/test-tool/README.md:
```diff
@@ -0,0 +1,58 @@
+
+## commands
+
+* The `init` command can be used to create a default configuration file.
```
it's better to add the  whole command samples into the doc. so we can directly see the command clearly.

### Step 1: Initialize in current directory
test-tool init --home .

### Step 2: Prepare test environment
test-tool prepare --home .

### Step 3: Start the test tool
test-tool start --home .

### Step 4: Run batch tests
test-tool batch-test depositLST --home .

### Query transactions for specific batch and status
test-tool query-tx-record <msgType> <batch-id> <status>

TimmyExogenous replied:
it will be added in the next commit.

TimmyExogenous committed:
add more code comment
fix the issue of incorrect error return when checking delegations
adjust the opAmount when batch sending delegations
fix the lint error in readme
add IntervalAfterDelegations and more code comments
print txID for debug
add staker funding for the automated test
fix the issue from code comment
rename test-tool to exocore-test-tool and add more details in the readme

TimmyExogenous commented on cmd/test-tool/main.go:
```diff
@@ -0,0 +1,316 @@
+
+		// Serialize the default configuration to TOML format
+		encoder := toml.NewEncoder(file)
+		if err := encoder.Encode(batch.DefaultTestToolConfig); err != nil {
+			fmt.Printf("failed to encode config to TOML: %s\r\n", err)
+		}
```
It will be added in the next commit.

TimmyExogenous commented on cmd/exocore-test-tool/main.go:
```diff
@@ -0,0 +1,317 @@
+	Args:    cobra.NoArgs,
+	Run: func(_ *cobra.Command, _ []string) {
+		// Start the app manager in a separate goroutine
+		go func() {
+			if err := appManager.Start(); err != nil {
+				fmt.Printf("Error starting the test tool: %s\r\n", err)
+			}
+		}()
```
dismiss it, because it's only used in the test tool

TimmyExogenous commented on cmd/exocore-test-tool/main.go:
```diff
@@ -0,0 +1,317 @@
+	Args:    cobra.ExactArgs(1),
+	Run: func(_ *cobra.Command, args []string) {
+		// Start the app manager in a separate goroutine
+		go func() {
+			if err := appManager.ExecuteBatchTestForType(args[0]); err != nil {
+				fmt.Printf("failed to execute the batch test for the specified message type,"+
+					"msgType:%s,err: %s\r\n", args[0], err)
+			}
+		}()
```
dismiss it, because it's only used in the test tool

TimmyExogenous commented on testutil/batch/send_txs.go:
```diff
@@ -0,0 +1,310 @@
+		}
+
+		// Log a message indicating the receipt is not yet available
+		logger.Info("can't get the receipt of EVM transaction, continue waiting", "waitDuration", waitDuration, "err", err)
```
It will be initialized globally.

MaxMustermann2 commented on precompiles/avs/tx.go:
```diff
@@ -49,6 +49,7 @@ func (p Precompile) RegisterAVS(
 	// The AVS registration is done by the calling contract.
 	avsParams.AvsAddress = contract.CallerAddress
 	avsParams.Action = avstypes.RegisterAction
+	fmt.Println("call RegisterAVS, the avs addr is:", avsParams.AvsAddress)
```
Remove debug

TimmyExogenous replied:
thanks, it will be fixed in the next commit.

MaxMustermann2 commented on precompiles/assets/assets_binding.go:
```diff
@@ -0,0 +1,396 @@
+// Code generated - DO NOT EDIT.
```
Can you document how to regenerate this?

TimmyExogenous replied:
Yes, it will be added in the next commit.

bwhour reviewed (approved):
looks good.
